### PR TITLE
Location component

### PR DESCRIPTION
**Pull request by philippjfr**
Massive redesign of the original Location component written by @MarcSkovMadsen. The problem is that it's not really a Viewable component and therefore should not be a widget. It's also weird adding it to a layout if it doesn't actually display anything. Therefore I introduced non-viewable baseclasses for Panel components and the location is automatically added as a Document root and then made accessible via `pn.state.location`.

In order to make the Location component work without also being a Viewable objects the baseclasses are now organized as such:

* `Layoutable(Parameterized)`: Defines parameters concerned with layout and style
* `ServableMixin()`: Mixin class that defines methods to serve object on a server
* `Renderable(Parameterized)`: Defines methods that allow the object to rendered to a bokeh model
* `Viewable(Renderable, Layoutable, ServableMixin)`: Defines methods that make the object viewable in a notebook, save it and serve it
* `Syncable(Renderable)`: Defines methods that allow parameters to be synced with bokeh model properties
* `Reactive(Viewable, Syncable)`: A syncable viewable that defines various methods to link it to other objects.

Supersedes https://github.com/holoviz/panel/pull/1101

### CODE REVIEW

**Comment by philippjfr** [comment]
Was hoping to get this into 0.9.0 after all but it touches too many internals for me to safely include it.

File: .gitignore [hunk 1]
```diff
@@ -123,4 +123,4 @@ package-lock.json
 # vscode settings
 .vscode/
 .vscode/panel.code-workspace
-discover/
+discover/
```
Comment (jbednar) [inline .gitignore]:
```suggestion
discover/

```

File: examples/user_guide/Deploy_and_Export.ipynb [hunk 2]
```diff
@@ -332,6 +327,42 @@
     "\n",
     "This mechanism may be used to modify the behavior of an app dependending on parameters provided in the URL. \n",
     "\n",
+    "#### Cookies\n",
+    "\n",
+    "The `panel.state.cookies` will allow accessing the cookies stored in the browser and on the bokeh server.\n",
```
Comment (jbednar) [inline examples/user_guide/Deploy_and_Export.ipynb]:
```suggestion
    "`pn.state.cookies` allows accessing the cookies stored in the browser and on the bokeh server.\n",
```

File: examples/user_guide/Deploy_and_Export.ipynb [hunk 3]
```diff
@@ -332,6 +327,42 @@
...
+    "\n",
+    "#### Headers\n",
+    "\n",
+    "The `panel.state.headers` will allow accessing the HTTP headers stored in the browser and on the bokeh server.\n",
```
Comment (jbednar) [inline examples/user_guide/Deploy_and_Export.ipynb]:
```suggestion
    "`pn.state.headers` allows accessing the HTTP headers stored in the browser and on the bokeh server.\n",
```

File: examples/user_guide/Deploy_and_Export.ipynb [hunk 4]
```diff
@@ -332,6 +327,42 @@
     "\n",
     "This mechanism may be used to modify the behavior of an app dependending on parameters provided in the URL. \n",
```
Comment (jbednar) [inline examples/user_guide/Deploy_and_Export.ipynb]:
```suggestion
    "This mechanism may be used to modify the behavior of an app depending on parameters provided in the URL. \n",
```

File: examples/user_guide/Deploy_and_Export.ipynb [hunk 5]
```diff
@@ -332,6 +327,42 @@
...
+    "\n",
+    "The `panel.state.cookies` will allow accessing the cookies stored in the browser and on the bokeh server.\n",
+    "\n",
+    "#### Headers\n",
+    "\n",
+    "The `panel.state.headers` will allow accessing the HTTP headers stored in the browser and on the bokeh server.\n",
+    "\n",
+    "#### Location\n",
+    "\n",
+    "When starting a server session Panel will attach a `Location` component which can be accessed using `pn.state.location`. The `Location` component servers a number of functions:\n",
```
Comment (jbednar) [inline examples/user_guide/Deploy_and_Export.ipynb]:
```suggestion
    "When starting a server session Panel will attach a `Location` component, which can be accessed using `pn.state.location`. The `Location` component serves a number of functions:\n",
```

File: examples/user_guide/Deploy_and_Export.ipynb [hunk 6]
```diff
@@ -332,6 +327,42 @@
...
+    "\n",
+    "- Navigation between pages via ``pathname``\n",
+    "- Sharing (parts of) the page state in the url as ``search`` parameters for bookmarking and sharing.\n",
+    "- Navigating to subsections of the page via the ``hash_`` parameter.\n",
+    "\n",
+    "##### Core\n",
+    "\n",
+    "* **``pathname``** (string): pathname part of the url, e.g. '/user_guide/Interact.html'.\n",
+    "* **``search``** (string): search part of the url e.g. '?color=blue'.\n",
```
Comment (jbednar) [inline examples/user_guide/Deploy_and_Export.ipynb]:
```suggestion
    "* **``search``** (string): search part of the url, e.g. '?color=blue'.\n",
```

File: panel/io/server.py [hunk 7]
```diff
@@ -131,15 +131,19 @@ def serve(panels, port=0, websocket_origin=None, loop=None, show=True,
       An HTML title for the application
     verbose: boolean (optional, default=True)
       Whether to print the address and port
+    location : boolean or panel.io.location.Location
+      Whether to create a Location component to observe and
+      set the URL location.
```
Comment (jbednar) [inline panel/io/server.py]: Is there some legitimate reason not to create such a component?  If so, explain.

File: panel/io/server.py [hunk 8]
```diff
@@ -168,6 +172,9 @@ def get_server(panel, port=0, websocket_origin=None, loop=None,
       An HTML title for the application
     verbose: boolean (optional, default=False)
       Whether to report the address and port
+    location : boolean or panel.io.location.Location
+      Whether to create a Location component to observe and
+      set the URL location.
```
Comment (jbednar) [inline panel/io/server.py]: Is there some legitimate reason not to create such a component? If so, explain.

File: panel/io/state.py [hunk 9]
```diff
@@ -20,16 +20,8 @@ class _state(param.Parameterized):
...
-       Global location you can use to cache large datasets or expensive computation results
-       across multiple client sessions for a given server.""") 
-
     webdriver = param.Parameter(default=None, doc="""
-        Selenium webdriver used to export bokeh models to pngs.""")
-
-    _curdoc = param.ClassSelector(class_=Document, doc="""
-        The bokeh Document for which a server event is currently being
-        processed.""")
+      Selenium webdriver used to export bokeh models to pngs.""")
```
Comment (jbednar) [inline panel/io/state.py]:
```suggestion
      Selenium webdriver used to export bokeh models to PNG images.""")
```

File: panel/models/location.py [hunk 10]
```diff
@@ -0,0 +1,41 @@
+"""This module provides a Bokeh Location Model as a wrapper around the JS window.location api"""
```
Comment (jbednar) [inline panel/models/location.py]:
```suggestion
"""This module provides a Bokeh Location Model as a wrapper around the JS window.location API"""
```

File: panel/models/location.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+from bokeh.core.properties import Bool, String
+from bokeh.models import Model
+
+
+class Location(Model):
+    """
+    A python wrapper around the JS `window.location` api. See
```
Comment (jbednar) [inline panel/models/location.py]:
```suggestion
    A python wrapper around the JS `window.location` API. See
```

File: panel/viewable.py [hunk 12]
```diff
@@ -311,45 +319,26 @@ def show(self, title=None, port=0, websocket_origin=None, threaded=False,
...
-    displays itself by transforming the object(s) being wrapped into
-    models that can be served using bokeh's layout engine. The class
-    also defines various methods that allow Viewable objects to be
-    displayed in the notebook and on bokeh server.
-    """
-
-    __abstract = True
 
-    _preprocessing_hooks = []
+class Renderable(param.Parameterized):
```
Comment (jbednar) [inline panel/viewable.py]: Seems like each abstract base class should have a class docstring explaining what the function of the class is.

File: panel/viewable.py [hunk 13]
```diff
@@ -1096,3 +1085,53 @@ def jslink(self, target, code=None, args=None, bidirectional=False, **links):
         from .links import Link
         return Link(self, target, properties=links, code=code, args=args,
                     bidirectional=bidirectional)
+
+
+class Reactive(Syncable, Viewable):
+    """
+    Reactive is a Viewable object that also supports syncing between
+    the objects parameters and the underlying bokeh model either via
```
Comment (jbednar) [inline panel/viewable.py]:
```suggestion
    the object's parameters and the underlying bokeh model either via
```

**Pull request merged**